File: conway.erl
-module(conway).
-export([demo_grid/0, run_simulation/2]).

-record(grid, {height=10, width=10, rows}).
-record(transition, {y, x, state}).

mod(X, Y) when X > 0 -> X rem Y;
mod(X, Y) when X < 0 -> Y + X rem Y;
mod(0, _Y) -> 0.

new(Height, Width) ->
    Column = array:new(Height),
    Array2D = array:map(fun(_X, _T) -> array:new([{size, Width}, {fixed, true}, {default, empty}]) end, Column),
    #grid{height=Height, width=Width, rows=Array2D}.

query(Grid, Y, X) ->
    Row = array:get(mod(Y, Grid#grid.height), Grid#grid.rows),
    array:get(mod(X, Grid#grid.width), Row).

assign(Grid, Y, X, State) ->
    Row =  array:get(mod(Y, Grid#grid.height), Grid#grid.rows),
    UpdatedRow = array:set(mod(X, Grid#grid.width), State, Row),
    Rows = array:set(mod(Y, Grid#grid.height), UpdatedRow, Grid#grid.rows),
    Grid#grid{rows=Rows}.

draw(Value) ->
    case Value of
        empty -> "-";
        alive -> "*"
    end.

count_neighbors(Grid, Y, X) ->
    N  = query(Grid, Y + 1, X    ),
    NE = query(Grid, Y + 1, X + 1),
    E  = query(Grid, Y    , X + 1),
    SE = query(Grid, Y - 1, X + 1),
    S  = query(Grid, Y - 1, X    ),
    SW = query(Grid, Y - 1, X - 1),
    W  = query(Grid, Y    , X - 1),
    NW = query(Grid, Y + 1, X - 1),
    NeighborStates = [N, NE, E, SE, S, SW, W, NW],
    length(lists:filter(fun(State) -> State == alive end, NeighborStates)).

game_logic(State, Neighbors) ->
    case State of
        alive -> if
            Neighbors < 2 -> empty;
            Neighbors > 3 -> empty;
            true -> State end;
        empty -> if
            Neighbors == 3 -> alive;
            true -> State end
    end.

step_cell(Grid, Y, X) ->
    State = query(Grid, Y, X),
    Neighbors = count_neighbors(Grid, Y, X),
    NextState = game_logic(State, Neighbors),
    #transition{y=Y, x=X, state=NextState}.

simulate(Grid) ->
    Rows = array:map(fun(Y, Row) ->
        array:map(fun(X, _CellValue) ->
            Transition = step_cell(Grid, Y, X),
            Transition#transition.state end, Row) end, Grid#grid.rows),
    Grid#grid{rows=Rows}.

print(Grid) ->
    ColumnIndex = lists:seq(0, Grid#grid.width - 1),
    RowIndex = lists:seq(0, Grid#grid.height - 1),
    lists:foreach(fun(Y) ->
        io:format("|"),
        lists:foreach(fun(X) ->
            io:format(" ~s |", [draw(query(Grid, Y, X))]) end, ColumnIndex),
            io:format("~n") end, RowIndex).

demo_grid() ->
    Grid0 = new(10, 20),
    Grid1 = assign(Grid0, 0, 3, alive),
    Grid2 = assign(Grid1, 1, 4, alive),
    Grid3 = assign(Grid2, 2, 2, alive),
    Grid4 = assign(Grid3, 2, 3, alive),
    assign(Grid4, 2, 4, alive).

clear_screen() ->
    io:format("\ec").

end_simulation() ->
    io:format("~nEnd of simulation~n").

run_simulation(Grid, N) ->
    clear_screen(),
    UpdatedGrid = simulate(Grid),
    print(UpdatedGrid),
    timer:sleep(500),
    case N > 0 of
        true -> run_simulation(UpdatedGrid, N - 1);
        false -> end_simulation()
    end.

File: Matlab/conway.m
clc, clearvars

function out = modulo(x, y)
    if (x > 0)
        out = mod(x, y);
    elseif (x < 0)
        out = mod(y + x, y);
    else
        out = 0;
    endif
end

function out = query(c, x, y)
    width = length(c);
    height = length(c(:,1));
    out = c(modulo(x, height) + 1, modulo(y, width) + 1);
end

function c = assign(c, x, y, state)
    width = length(c);
    height = length(c(:,1));
   c(modulo(x, height) + 1, modulo(y, width) + 1) = state;
end

function neighbors = count_neighbors(c, x, y)
    n  = query(c, x + 1, y    );
    ne = query(c, x + 1, y + 1);
    e  = query(c, x,     y + 1);
    se = query(c, x - 1, y + 1);
    s  = query(c, x - 1, y    );
    sw = query(c, x - 1, y - 1);
    w  = query(c, x,     y - 1);
    nw = query(c, x + 1, y - 1);
    neighbors = n + ne + e + se + s + sw + w + nw;
end

function next = game_logic(state, neighbors)
    if (state == 1)
        if (neighbors < 2)
            next = 0;
        elseif (neighbors > 3)
            next = 0;
        else
            next = 1;
        endif
    else
        if (neighbors == 3)
            next = 1;
        else
            next = 0;
        endif
    endif
end

function next = step_cell(c, x, y)
    state = query(c, x, y);
    neighbors = count_neighbors(c, x, y);
    next = game_logic(state, neighbors);
end

function new_c = simulate(c)
    width = length(c);
    height = length(c(:,1));
    new_c = zeros(height, width);
    for i = 1:height
        for j = 1:width
            next = step_cell(c, i, j);
            new_c = assign(new_c, i, j, next);
        endfor
    endfor
end

C = zeros(10, 20);
C = assign(C, 1, 4, 1);
C = assign(C, 2, 5, 1);
C = assign(C, 3, 3, 1);
C = assign(C, 3, 4, 1);
C = assign(C, 3, 5, 1)

while (true)
    % clc
    C = simulate(C)
    pause(0.1)
endwhile

% octave conway.m
